### PR TITLE
Cross validation alpha

### PR DESCRIPTION
**Pull request by AliHashemi-ai**
Hi @agramfort , 
could you please take a look at this PR for implementing spatial CV techniques to tune alpha based on MSE using sklearn.

### CODE REVIEW

**Review by agramfort — COMMENTED** [review]
Did you already test it ? Does it work as expected? Also check that the score used in what you expect eg MSE or R2?

**Comment by AliHashemi-ai** [comment]
> Did you already test it ? Does it work as expected? Also check that the score used in what you expect eg MSE or R2?

Yes I have already tested it and all the tests have passed. The score that I used in my papers is based on MSE. 
@agramfort  do you have an idea how we can do it for temporal CV? do the cross validation for time samples?

**Comment by agramfort** [comment]
Can you add a test for this new code? I would test that it selects an alpha that works well eg identifies the support. I would test this only for a type 1 or type 2 method.

**Comment by AliHashemi-ai** [comment]
> @AliHashemi-ai you cannot add code here without testing it with the continuous integration. Like we did with the test_ function for estimators.

I see your point @agramfort . I am not sure how should I implement this. Currently, we have this 
`@pytest.mark.parametrize(
    "solver,alpha,rtol,atol,cov_type,extra_params",
    [
        (iterative_L1, 0.1, 1e-1, 5e-1, "diag", {}),`
that parameterize different solvers. 

Now, should I add another line to this to have run_spatial_cv(solver, L, y, cov=None): ? but then it needs to have solver as an input?

**Comment by agramfort** [comment]
yes add a new function test_run_spatial_cv that tests the code with some
solver.

**Comment by AliHashemi-ai** [comment]
Hi @agramfort  could you please check the latest updates for testing the spatial cv and merge it to the master if you confirm it. We also discuss this in our meeting today.

File: bsi_zoo/estimators.py [hunk 1]
```diff
@@ -63,6 +68,76 @@ def _solve_reweighted_lasso(
...
+            param_grid=dict(alpha=self.alphas),
+            scoring="neg_mean_squared_error",
+            cv=5,
+            n_jobs=-1,
+        )
+        clf.fit(L, y)
+        self.alpha = clf.best_estimator_.alpha
+        return self
+
+    def predict(self, y):
```
Comment (agramfort) [inline bsi_zoo/estimators.py:113]: To me alpha is a function of y so grid search should be done in the predict

File: bsi_zoo/estimators.py [hunk 2]
```diff
@@ -63,6 +68,76 @@ def _solve_reweighted_lasso(
...
+        if self.cov is None:
+            self.coef_ = self.solver(self.L_, y, alpha=self.alpha)
+        else:
+            self.coef_ = self.solver(self.L_, y, self.cov, alpha=self.alpha)
+
+        return self.coef_
+
+
+class Spatial_CV_Solver(BaseEstimator, ClassifierMixin):
```
Comment (agramfort) [inline bsi_zoo/estimators.py:92]:
```suggestion
class SpatialCVSolver(BaseEstimator, ClassifierMixin):
```

File: bsi_zoo/estimators.py [hunk 3]
```diff
@@ -63,6 +68,76 @@ def _solve_reweighted_lasso(
...
+    def __init__(self, solver, alpha=None, alphas = np.linspace(1.4, 0.1, 20), cov=None):
```
Comment (agramfort) [inline bsi_zoo/estimators.py:93]: Run flake8

File: bsi_zoo/estimators.py [hunk 4]
```diff
@@ -63,6 +68,76 @@ def _solve_reweighted_lasso(
...
+    def predict(self, y):
+        if self.cov is None:
+            self.coef_ = self.solver(self.L_, y, alpha=self.alpha)
+        else:
+            self.coef_ = self.solver(self.L_, y, self.cov, alpha=self.alpha)
+
+        return self.coef_
+
+
+def run_spatial_cv(solver, L, y, cov=None, n_orient=1, **extra_params):
```
Comment (agramfort) [inline bsi_zoo/estimators.py:122]: This function is now redundant with the class. Either the class uses this function or we remove it

File: bsi_zoo/tests/test_run_spatial_cv.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        (iterative_L2_typeII, 0.1, 1e-1, 5e-1, "full", {}),
+        (gamma_map, 0.2, 1e-1, 5e-1, "full", {"update_mode": 1}),
+        (gamma_map, 0.2, 1e-1, 5e-1, "full", {"update_mode": 2}),
+        (gamma_map, 0.2, 1e-1, 5e-1, "full", {"update_mode": 3}),
+    ],
+)
+def test_run_spatial_cv(
+    n_times,
+    solver,
+    alpha,
```
Comment (agramfort) [inline bsi_zoo/tests/test_run_spatial_cv.py:40]: Why do you need alpha here ?

File: bsi_zoo/tests/test_run_spatial_cv.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    orientation_type,
+    extra_params,
+    save_estimates=False,
+):
+
+    path_to_leadfield = get_leadfield_path(
+        subject, orientation_type
+    )  # setting leadfield paths
+
+    n_orient = 1 if orientation_type == "fixed" else 3
```
Comment (agramfort) [inline bsi_zoo/tests/test_run_spatial_cv.py:55]: This test should not necessitate to use stc objects. I would just check that coef_ is the solver is ok in some sense

**Comment by agramfort** [comment]
@AliHashemi-ai this is green now. HOwever the Temporal CV is not doing the right thing for now. Can you have a look?

**Comment by AliHashemi-ai** [comment]
> @AliHashemi-ai this is green now. HOwever the Temporal CV is not doing the right thing for now. Can you have a look?

Hi @agramfort  I have added the temporal CV to the code based on the Bregman logdet divergence

**Comment by agramfort** [comment]
@AliHashemi-ai the CI is red. Can you look into it? I will review and edit afterwards. Thanks 🙏

File: bsi_zoo/tests/test_cross_val.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    atol,
+    cov_type,
+    nnz,
+    orientation_type,
+    extra_params,
+):
+    assert cv_type in ["temporal", "spatial"]
+
+    if estimator in [iterative_L1_typeII, iterative_L2_typeII] and cov_type == "full":
+        pytest.skip("iterative L1 type 2 breaks with full covariance")
```
Comment (agramfort) [inline bsi_zoo/tests/test_cross_val.py:50]: @AliHashemi-ai when the cov is full you cannot reduce the number of channels without changing the cov attribute/parameter. How did you deal with this in the past?

File: bsi_zoo/cross_val.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
+from scipy import linalg
+import numpy as np
+from sklearn.model_selection import GridSearchCV
+from sklearn.base import BaseEstimator, ClassifierMixin, clone
+from sklearn.model_selection import check_cv
```
Comment (agramfort) [inline bsi_zoo/cross_val.py:5]: @AliHashemi-ai I moved all the cross-val code to one file

File: bsi_zoo/cross_val.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+def logdet_bregman_div_distance_nll(y, Sigma_Y):
+    """Compute the Gaussian log likelihood of y given covariance Sigma_Y."""
+    # compute empirical covariance of the test set
+    precision = np.inv(Sigma_Y)
+    _, n_features = y.shape
+    log_like = -.5 * (y * (np.dot(y, precision))).sum(axis=1)
+    log_like -= .5 * (n_features * np.log(2. * np.pi) - _logdet(precision))
+    out = np.mean(log_like)
+    return out
```
Comment (agramfort) [inline bsi_zoo/cross_val.py:46]: @AliHashemi-ai this comes from mne. Any reason we need something different?

File: bsi_zoo/cross_val.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+
+# def logdet_bregman_div_distance_nll(y, Sigma_Y):
+#     """Compute the log-det Bregman divergence between
+#     two matrices based on the calculation of Gaussian
+#     negative log-likelihood.
+#     """
+#     log_lik = _gaussian_loglik_scorer(y, Sigma_Y)
+#     nll = -2 * log_lik
+#     return nll
```
Comment (agramfort) [inline bsi_zoo/cross_val.py:85]: do we need the commented code above?

File: bsi_zoo/cross_val.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+            solver = clone(base_solver)
+            solver.set_params(alpha=alpha)
+            temporal_cv_scores = []
+            for train_idx, test_idx in cv.split(y.T):
+                solver.fit(self.L_, y[:, train_idx])
+                y_test = y[:, test_idx]
+                X_diag = np.sum(np.abs(solver.coef_), axis=1) != 0
+                # XXX this next formula seems to be wrong as it ignores
+                # the estimated or empirical source variance.
+                Sigma_Y = self.cov + (self.L_ * X_diag[None, :]) @ self.L_.T
```
Comment (agramfort) [inline bsi_zoo/cross_val.py:183]: @AliHashemi-ai this seems very broken for now. You should need here the estimated gamma or compute an estimate of the source variance. What do you think?

**Review by agramfort — COMMENTED** [review]
@AliHashemi-ai @anujanegi we can discuss this next time we talk. Please look at the code I wrote in the last couple of hours.

File: bsi_zoo/cross_val.py (hunk 11 above)
Comment (AliHashemi-ai) [reply bsi_zoo/cross_val.py:183]: @agramfort  You are correct. I need to replace X_diag with the an estimate of the source variance. I will fix it.

File: bsi_zoo/cross_val.py (hunk 9 above)
Comment (AliHashemi-ai) [reply bsi_zoo/cross_val.py:46]: @agramfort Initially, I indented to use the MNE code to implement the temporal cross validation metric since the Bregman logdet divergence can be reformulated as negative likelihood loss plus some constants. And since NNL is efficiently implemented in MNE, I wanted to use the code to do it efficiently. But, there was some difficulties that I decided to write the cv cross-validation metric from scratch.

File: bsi_zoo/cross_val.py (hunk 10 above)
Comment (AliHashemi-ai) [reply bsi_zoo/cross_val.py:85]: No I will remove that. (I explained the reason above)

File: bsi_zoo/cross_val.py (hunk 8 above)
Comment (AliHashemi-ai) [reply bsi_zoo/cross_val.py:5]: Sounds great Thank you

**Comment by AliHashemi-ai** [comment]
@agramfort I fixed the error in temporal cv for the iterative sqrt now it is green now could you please review it for the merge

**Comment by agramfort** [comment]
@AliHashemi-ai I just pushed a commit. Let me know what you think.

File: bsi_zoo/cross_val.py [hunk 12]
```diff
@@ -139,13 +121,15 @@ def _get_alpha(self, y):
...
                 solver.fit(self.L_, y[:, train_idx])
                 y_test = y[:, test_idx]
-                # X_diag = np.sum(np.abs(solver.coef_), axis=1) != 0
-                # Cov_X = np.cov(solver.coef_)
-                X_Sqaure = solver.coef_ @ solver.coef_.T
-                Cov_X = np.diag(np.linalg.norm(X_Sqaure, axis=0))
-                Sigma_Y = self.cov + ((self.L_ @ Cov_X) @ self.L_.T)
+
+                # X_Sqaure = solver.coef_ @ solver.coef_.T
+                # Cov_X = np.diag(np.linalg.norm(X_Sqaure, axis=0))
```
Comment (agramfort) [inline bsi_zoo/cross_val.py:125]: @AliHashemi-ai I don’t understand what you computed before.

**Comment by AliHashemi-ai** [comment]
Hi @agramfort  could you please check this PR and merge it if you confirm the changes? 
Currently the temporal CV does not work for sqrt estimator?
